packages/muelu/src/Graph/StructuredAggregation/MueLu_AggregationStructuredAlgorithm_def.hpp

```diff
@@ -134,11 +138,184 @@ namespace MueLu {
+    geoData->getGhostedNodesData(graph.GetDomainMap(), ghostedCoarseNodeCoarseLIDs,
+                                 ghostedCoarseNodeCoarsePIDs, ghostedCoarseNodeCoarseGIDs);
+
+    LO iGhosted, jGhosted, kGhosted, iCoarse, jCoarse, kCoarse, iRem, jRem, kRem;
+    LO ghostedCoarseNodeCoarseLID, aggId, rate;
+    LO dim = geoData->getNumDimensions();
+    LO rowCount = 0;
+    GO myGID;
+    bool iIsCoarse, jIsCoarse, kIsCoarse;
+    Array<LO> offsets = geoData->getOffsets();
+    Array<LO> colIndex(geoData->getNumLocalFineNodes()*std::pow(2,dim));
+    Array<LO> rowPtr(geoData->getNumLocalFineNodes()+1);
+    rowPtr[0] = 0;
+    ArrayRCP<size_t> nnzOnRow(geoData->getNumLocalFineNodes());
+
+    *out << "Loop over fine nodes and assign them to an aggregate and a rank" << std::endl;
+    for(LO nodeIdx = 0; nodeIdx < geoData->getNumLocalFineNodes(); ++nodeIdx) {
+      // Compute coarse ID associated with fine LID
+      geoData->getFineNodeGhostedTuple(nodeIdx, iGhosted, jGhosted, kGhosted);
+      iCoarse = iGhosted / geoData->getCoarseningRate(0);
+      iRem    = iGhosted % geoData->getCoarseningRate(0);
+      if(iGhosted - geoData->getOffset(0)
+         < geoData->getLocalFineNodesInDir(0) - geoData->getCoarseningEndRate(0)) {
+        rate = geoData->getCoarseningRate(0);
+      } else {
+        rate = geoData->getCoarseningEndRate(0);
+      }
+      if(iRem > (rate / 2)) { ++iCoarse; }
+      if(coupled && (geoData->getStartGhostedCoarseNode(0)*geoData->getCoarseningRate(0)
+                     > geoData->getStartIndex(0))) { --iCoarse; }
+      jCoarse = jGhosted / geoData->getCoarseningRate(1);
+      jRem    = jGhosted % geoData->getCoarseningRate(1);
+      if(jGhosted - geoData->getOffset(1)
+         < geoData->getLocalFineNodesInDir(1) - geoData->getCoarseningEndRate(1)) {
+        rate = geoData->getCoarseningRate(1);
+      } else {
+        rate = geoData->getCoarseningEndRate(1);
+      }
+      if(jRem > (rate / 2)) { ++jCoarse; }
+      if(coupled && (geoData->getStartGhostedCoarseNode(1)*geoData->getCoarseningRate(1)
+                     > geoData->getStartIndex(1))) { --jCoarse; }
+      kCoarse = kGhosted / geoData->getCoarseningRate(2);
+      kRem    = kGhosted % geoData->getCoarseningRate(2);
+      if(kGhosted - geoData->getOffset(2)
+         < geoData->getLocalFineNodesInDir(2) - geoData->getCoarseningEndRate(2)) {
+        rate = geoData->getCoarseningRate(2);
+      } else {
+        rate = geoData->getCoarseningEndRate(2);
+      }
+      if(kRem > (rate / 2)) { ++kCoarse; }
+      if(coupled && (geoData->getStartGhostedCoarseNode(2)*geoData->getCoarseningRate(2)
+                     > geoData->getStartIndex(2))) { --kCoarse; }
+      geoData->getCoarseNodeGhostedLID(iCoarse, jCoarse, kCoarse, ghostedCoarseNodeCoarseLID);
+
+      // Fill Graph
+      // Check if Fine node lies on Coarse Node
+      iIsCoarse = false; jIsCoarse = false; kIsCoarse = false;
```

Comment:
replace `iIsCoarse`, `jIsCoarse` and `kIsCoarse` by an array and loop over `numDimensions` to fill said array